java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java

```diff
@@ -0,0 +1,876 @@
+          || (indicesShape.length == 0)
+          || (indicesShape[0] != numNonZero)) {
+        throw new IllegalArgumentException(
+            "Invalid indices shape, expected [numNonZero, dimension] or [numNonZero] found "
+                + Arrays.toString(indicesShape));
+      }
+      long elementCount = OrtUtil.elementCount(indicesShape);
+      if (elementCount != indices.remaining()) {
+        throw new IllegalArgumentException(
+            "Unexpected number of indices found in buffer, expected "
+                + elementCount
+                + " found "
+                + indices.remaining());
+      }
+      this.indicesShape = indicesShape;
+      if (data.remaining() != numNonZero) {
+        throw new IllegalArgumentException(
+            "Expected data.remaining() - "
+                + data.remaining()
+                + " to equal numNonZero - "
+                + numNonZero);
+      }
+    }
+
+    @Override
+    public long[] getValuesShape() {
+      return new long[] {getNumNonZeroElements()};
+    }
+
+    @Override
+    public long[] getIndicesShape() {
+      return indicesShape;
+    }
+
+    @Override
+    public OnnxJavaType getIndicesType() {
+      return OnnxJavaType.INT64;
+    }
+
+    @Override
+    public SparseTensorType getSparsityType() {
+      return SparseTensorType.COO;
+    }
+  }
+
+  /** The Java side representation of a CSRC sparse tensor. */
+  public static final class CSRCTensor extends BaseSparseTensor<LongBuffer> {
+    private final LongBuffer innerIndices;
+
+    /**
+     * Creates a CSRC sparse tensor suitable for constructing an ORT Sparse Tensor.
+     *
+     * @param outerIndices The outer indices.
+     * @param innerIndices The inner indices.
+     * @param data The data.
+     * @param shape The dense shape.
+     * @param type The data type.
+     * @param numNonZero The number of non-zero elements.
+     */
+    public CSRCTensor(
+        LongBuffer outerIndices,
+        LongBuffer innerIndices,
+        Buffer data,
+        long[] shape,
+        OnnxJavaType type,
+        long numNonZero) {
+      super(outerIndices, data, shape, type, numNonZero);
+      this.innerIndices = innerIndices;
+      long expectedRows = shape[0] + 1;
+      if (outerIndices.remaining() != expectedRows) {
+        throw new IllegalArgumentException(
+            "Outer indices should be equal to the number of rows + 1 in the dense shape, found "
+                + outerIndices.remaining()
+                + ", expected "
+                + expectedRows);
+      }
+      if (innerIndices.remaining() != numNonZero) {
+        throw new IllegalArgumentException(
+            "Inner indices should be equal to the number of non-zero elements, found "
+                + innerIndices.remaining()
+                + ", expected "
+                + numNonZero);
+      }
+    }
+
+    @Override
+    public long[] getValuesShape() {
+      return new long[] {getNumNonZeroElements()};
+    }
```

Comment:
Can common things be in a base class? Such as getValues(), getValuesShape(), getIndicesType(), getSparsityType()?

Reply:
Yes, they could be build on construction and passed into the super constructor for `BaseSparseTensor`. I'd need private static methods to compute the various shapes, which would be about as much code as the current way, but it would be slightly more memory efficient. One issue however is that arrays are mutable, so strictly speaking users can modify the array elements which would confuse that sparse tensor. Usually we'd defensively copy them on the accessor which removes the memory efficiency.

Reply:
I see. I will leave it up to you, but your comment to that effect would be great.

Reply:
I've moved most of those methods onto the base class. `getSparsityType` is still on the subclasses, I find it a little easier to read there as it's a constant for those subtypes. Makes it harder to get confused.